Bugfix: Fighters should find new parents

samrocketman opened the pull request:
Bug description
---------------

Fighters hang if the flagship has fighter bays available.  The fighters will race to the flagship and join.  If the flagship fills up with bays the fighters do not try to find a new parent.  Even if escorts have free bays the fighter will not try to find a new bay to join.

Expected behavior
-----------------

After a ship's bays are filled the fighter should try to find a new parent among player escorts.  If an escort has an available bay then the fighter (or drone) should board that escort.

If there are no bays among escorts then the abandoned fighter should fall back to keeping station with the player flagship.

Actual behavior
---------------

Fighters race to flagship.  When bays are filled the fighters keep station with the flagship.  Even if bays are available among escorts the fighter or drone will not attempt to board amongst a new parent.

Save game
---------

You can test with this [save game shared in discord][discord].

Testing
-------

Tested the following scenarios where a player fleet attempted to pick up two (2) abandoned fighters with various fleet configurations with free bays.

| Abandoned fighters | Flagship Free Bays | Escort Free Bays |
| ------------------ | ------------------ | ---------------- |
| 2 | 2 | 0 |
|                  2 |                  1 |                6 |
|                  2 |                  1 |                0 |
|                  2 |                  1 |                1 |
|                  2 |                  0 |                2 |
|                  2 |                  0 |                1 |
|                  2 |                  0 |                0 |

- In all above tested scenarios, if the flagship has a bay the fighters would attempt to board.
- If the flagship was full or the flagship has no bays then the fighters would board an escort.
- If no bays are available then the abandoned fighters would keep station with the player flagship.

Performance Impact
------------------

None.  ~~The way `Ship::CanCarry` has been rewritten it should theoretically be more performant than the old method as the number of player escorts grows.~~

I reverted `CanCarry` changes.  Unless I get feedback to restore my changes I'll leave it.  I think my version is better performance but it has its own drawbacks.

[discord]: https://discordapp.com/channels/251118043411775489/266345072554016768/979048779724517389

samrocketman commented:
Posted a video in discord https://discordapp.com/channels/251118043411775489/266345072554016768/979119391532798042

samrocketman commented:
https://discord.com/channels/251118043411775489/266345072554016768/979228330475794462 a review from the original person reporting this which caused me to look into it:

matthew.najmon says:

> samrocketman btw, got a chance to test out the new fighter docking behavior, and it works beautifully. much thanks, for a fix I've been wanting for years now.

samrocketman commented:
# Warning

After some play testing I've found this change to be unstable.  `government` segfaults (which I removed to see if the issue was fixed).  Upon removal then `HasBays()` segfaults.  I don't understand why this is segfaulting.  Perhaps someone with more experience could explain to me.

samrocketman commented:
Locally, I completely reverted `source/Ship.cpp` and still got a segfault.  I think this has to do with how I'm calling CanBeCarried in `source/AI.cpp` and Ship.cpp is just a symptom.

samrocketman commented:
# Fixed segfault

The segfault was indeed caused by AI. 90a54bce6d340b8c302250d884b70c096ec383d3 is the fix.

# Question

I went ahead and reverted CanCarry.  Should I use my original CanCarry implementation which I introduced in 0028fa43ed56ba24b2f52f06157bd81b67c38078 and later reverted?

Currently, this PR does not update CanCarry.  I feel like my version would be a little more performant but the draw back is ships may race to board escorts.

samrocketman commented:
I am not convinced this is stable, yet so am converting to a draft.

Hurleveur commented on source/AI.cpp:
```diff
@@ -829,7 +829,9 @@ void AI::Step(const PlayerInfo &player, Command &activeCommands)
 
-				it->SetParent(parent);
+				// Only set a new parent if the new parent can legitimately carry the fighter
+				if(!it->GetParent() || !parent || !parent->GetParent() || parent->CanCarry(*it))
```
why `|| !parent->GetParent()`? shouldn't it be an `&&`

samrocketman replied:
This is required for the abandoned fighter to keep station with player flagship.

With out this, you get the following: strange fighter behavior bouncing around to many escorts that can't carry the ship.

This enables the fighter to finally resolve to the player flagship carrier (which it can't board) when there are no free bays available for the fighter to board amongst the fleet.

This is the final fallback behavior in the "testing done" section.

samrocketman replied:
I can add a code comment for clarity around this so it is obvious post-merge.

samrocketman replied:
I think this is resolved with the rewrite.

quyykk reviewed (commented):
Your fix seems to revert behavior of the previous lines of code. For example, the behavior that fighters have the flagship as parent if there are no bays anymore, even if the flashing has no bays whatsoever.

I think the better way to solve this is to look at the code that is responsible for finding a new parent. Probably the reason why fighters don't seem to reparent is because of the random call that triggers on average every 30s, or something else (I haven't looked at that code in detail).

samrocketman commented:
> I think the better way to solve this is to look at the code that is responsible for finding a new parent. Probably the reason why fighters don't seem to reparent is because of the random call that triggers on average every 30s, or something else (I haven't looked at that code in detail).

I believe this section of code that I'm editing is that code which triggers roughly every 30s (inside random int).  I browsed the code extensively for parent boarding and this is what I found.  If you know a better section I'm open to looking at what you reference.

---

Can you expand on what you mean by me reverting previous behavior?  Are you referring to me reverting my own code within this PR or further back in the history.

---

IMO rehoming a fighter should be pretty immediate.  I found the delay in rehoming to be pretty frustrating.  There's no reason it should run every 30s to find a new parent.  If a fighter has a valid parent it will never trigger the logic path

quyykk commented:
> Can you expand on what you mean by me reverting previous behavior? Are you referring to me reverting my own code within this PR or further back in the history.

I mean that you're doing the opposite of the code a couple of lines above the SetParent call. Namely the CanCarry check.

A couple lines above there's a comment explaining that if there is no suitable parent it assigns the flagship, even if it can't carry. But then later on you changed it to not update the parent.

> There's no reason it should run every 30s to find a new parent.

Yeah I agree.

samrocketman commented:
> A couple lines above there's a comment explaining that if there is no suitable parent it assigns the flagship, even if it can't carry. But then later on you changed it to not update the parent

Ok I will check at a computer it should be more clear in an IDE.  I'll look at cleaning up the full logic path since it is hard to tell from phone.  Thanks

samrocketman commented:
I started from scratch based on what I've learned about performance tuning in #6726 .

@quyykk ready for a fresh review

samrocketman commented:
Something I learned about how random works in endless sky.  `!Random::Int(1800)` will roll a random number between 0 and 1800.  It has to be 0 for it to succeed.  It's not on average every 30s to search for new parent.  Instead, it is more accurate to say `1:1800` chance to actually re-home a fighter.

In practice this ends up being much longer than 30s.

quyykk commented on source/AI.cpp:
```diff
@@ -769,8 +769,11 @@ void AI::Step(const PlayerInfo &player, Command &activeCommands)
+			// NPCs may take 30 seconds or longer to find a new parent.  Player
+			// owned fighter shouldn't take more than a few seconds.
+			bool findNewParent = it->IsYours() ? !Random::Int(30) : !Random::Int(1800);
+			bool parentHasSpace = !findNewParent || (inParentSystem && parent->CanCarry(*it));
```
```suggestion
			bool parentHasSpace = (inParentSystem && parent->CanCarry(*it));
```

Not sure why you need that condition here. Also the name doesn't fit anymore.

samrocketman replied:
Because why waste the CPU cycles calculating when it will fail due to random int?  CanCarry iterates across all escorts and is an expensive operation.  By setting it to boolean true before bothering to calculate it you skip that expensive calculation.

It gets ignored one line later

samrocketman commented:
HasBays does not appear to work but CanCarry does.  CanCarry is more expensive since it iterates across all escorts.  I am going to look at why HasBays does not work.

samrocketman commented:
I have figured out the problem (since I am testing a few PRs there was a bug in another PR).

I plan to use CanCarry but only for player escorts.  NPC escorts should use HasBays to keep game performance good.

samrocketman commented on source/AI.cpp:
```diff
@@ -769,8 +769,13 @@ void AI::Step(const PlayerInfo &player, Command &activeCommands)
 			bool parentHasSpace = inParentSystem && parent->BaysFree(it->Attributes().Category());
-			if(!hasParent || (!inParentSystem && !it->JumpFuel()) || (!parentHasSpace && !Random::Int(1800)))
+			if(findNewParent && parentHasSpace)
+				parentHasSpace = parent->CanCarry(*it);
```
This expensive call is only called roughly twice a second.  Otherwise, the faster BaysFree call is used.  BaysFree is used to determine if this call should occur at all.

This enables fighters to choose different parents to return.  This faster path finding means from a player perspective abandoned fighters will in parallel choose a different parent which has free bays.

samrocketman reviewed (commented):
# :green_circle: Ready for review

- Abandoned fighters take less than a second to decide on where to board in your fleet if there's bays available.
- Performance issues fixed.  Faster `BaysFree` call is used most of the time.  Twice a second a new escort may be selected via `CanCarry` if there's no available bays for the returning fighter.  i.e. other fighters have already boarded or are about to board the carrier so choose another parent.
- Fighters choose different parents to return.  No racing back to the same parent if other escorts have bays available.

samrocketman commented:
Pushed a fix where if the player has a large number of excess fighters they do not scramble in circles.

samrocketman commented:
This PR solves some issues discussed in #6329

quyykk commented on source/Ship.cpp:
```diff
@@ -3467,9 +3467,14 @@ bool Ship::CanCarry(const Ship &ship) const
-			&& !escort->IsDestroyed())
+		if(!escort)
+			continue;
+		if(escort == ship.shared_from_this())
```
```suggestion
		if(escort.get() == &ship)
```

quyykk commented on source/Ship.cpp:
```diff
@@ -3467,9 +3467,14 @@ bool Ship::CanCarry(const Ship &ship) const
+		if(!escort)
+			continue;
+		if(escort == ship.shared_from_this())
+			break;
```
Why the break? If for some reason a ship has itself as an escort it should be ignored and the search should continue no?

quyykk commented on source/AI.cpp:
```diff
@@ -769,8 +769,13 @@ void AI::Step(const PlayerInfo &player, Command &activeCommands)
 			// A carried ship must belong to the same government as its parent to dock with it.
 			bool hasParent = parent && !parent->IsDestroyed() && parent->GetGovernment() == gov;
 			bool inParentSystem = hasParent && parent->GetSystem() == it->GetSystem();
+			// NPCs may take 30 seconds or longer to find a new parent.  Player
```
```suggestion
			// NPCs may take 30 seconds or longer to find a new parent. Player
```

quyykk commented on source/AI.cpp:
```diff
@@ -829,7 +834,13 @@ void AI::Step(const PlayerInfo &player, Command &activeCommands)
-				it->SetParent(parent);
+				// Player-owned carriables should defer to player carrier if
+				// selected parent can't carry it.
+				if(it->IsYours() && parent && parent->GetParent() && !parent->CanCarry(*it))
+					parent = parent->GetParent();
```
Do we need this? If the parent can't carry then it will trigger the big if statement above (twice a second) which would find a suitable carrier.

samrocketman replied:
@quyykk I added a comment so that its purpose is more clear.

Hurleveur commented on source/Ship.cpp:
```diff
@@ -3467,9 +3467,14 @@ bool Ship::CanCarry(const Ship &ship) const
+			continue;
+		if(escort == ship.shared_from_this())
+			break;
+		if(escort->attributes.Category() == category && !escort->IsDestroyed() && (!IsYours() || (IsYours() && escort->IsYours())))
```
```suggestion
		if(escort->attributes.Category() == category && !escort->IsDestroyed() &&
				(!IsYours() || (IsYours() && escort->IsYours())))
```
Convention is to limit line length to 100.

Hurleveur reviewed (dismissed):
Looking good.
One more step in the right direction for fighters to be better.

samrocketman commented:
I think I have addressed all of the feedback.  Also, this PR has been pretty thoroughly play tested at this point.  I haven't found anything else to fix.

RisingLeaf reviewed (approved):
Style checked and functionality seems correct. No playtest made

Amazinite commented on source/AI.cpp:
```diff
@@ -834,7 +839,15 @@ void AI::Step(const PlayerInfo &player, Command &activeCommands)
 
-				it->SetParent(parent);
+				// Player-owned carriables should defer to player carrier if
+				// selected parent can't carry it.  This is necessary to prevent
```
```suggestion
				// selected parent can't carry it. This is necessary to prevent
```
